test/cql-pytest/util.py

```diff
@@ -139,6 +143,53 @@ def new_secondary_index(cql, table, column, name='', extra=''):
     finally:
         cql.execute(f"DROP INDEX {keyspace}.{name}")
 
+# Helper function for establishing a connection with given username and password
+def cql_session(host, port, ssl, username, password):
+    profile = ExecutionProfile(
+        load_balancing_policy=RoundRobinPolicy(),
+        consistency_level=ConsistencyLevel.LOCAL_QUORUM,
+        serial_consistency_level=ConsistencyLevel.LOCAL_SERIAL,
+        # The default timeout (in seconds) for execute() commands is 10, which
+        # should have been more than enough, but in some extreme cases with a
+        # very slow debug build running on a very busy machine and a very slow
+        # request (e.g., a DROP KEYSPACE needing to drop multiple tables)
+        # 10 seconds may not be enough, so let's increase it. See issue #7838.
+        request_timeout = 120)
+    if ssl:
+        # Scylla does not support any earlier TLS protocol. If you try,
+        # you will get mysterious EOF errors (see issue #6971) :-(
+        ssl_context = ssl.SSLContext(ssl.PROTOCOL_TLSv1_2)
+    else:
+        ssl_context = None
+    cluster = Cluster(execution_profiles={EXEC_PROFILE_DEFAULT: profile},
+        contact_points=[host],
+        port=int(port),
+        # TODO: make the protocol version an option, to allow testing with
+        # different versions. If we drop this setting completely, it will
+        # mean pick the latest version supported by the client and the server.
+        protocol_version=4,
+        auth_provider=PlainTextAuthProvider(username=username, password=password),
+        ssl_context=ssl_context,
+    )
+    return cluster.connect()
+
+@contextmanager
+def new_user(cql, username=''):
+    if not username:
+        username = unique_name()
+    cql.execute(f"CREATE ROLE {username} WITH PASSWORD = '{username}' AND LOGIN = true")
+    try:
+        yield username
+    finally:
+        cql.execute(f"DROP ROLE {username}")
+
+@contextmanager
+def new_session(cql, username):
+    endpoint = cql.hosts[0].endpoint
+    session = cql_session(host=endpoint.address, port=endpoint.port, ssl=False, username=username, password=username)
+    yield session
+    session.shutdown()
```

Comment:
Is session.shutdown() correct? I thought it needs to be cluster.shutdown() (remember that session is what is returned by cluster.connect())... Or maybe session.shutdown() does cluster.shutdown(), or maybe session.shutdown() is enough?

Reply:
session.shutdown() shuts down the session (https://docs.datastax.com/en/drivers/python/3.2/api/cassandra/cluster.html#cassandra.cluster.Session.shutdown), while cluster.shutdown() shuts down all the sessions. I guess in this case I should make the whole helper, cql_session, a contextmanager and kill the cluster in there. I'll rework it and send v2. Technically it probably still "just works", but it may leave some stray useless control connections laying around in memory, so it's better to be thorough.